packages/atomic/project.json

```diff
@@ -1 +1,264 @@
-{}
+{
+  "namedInputs": {
+    "tests": ["{projectRoot}/src/**/*.{spec.{ts,tsx},spec.{ts,tsx}.snap}"],
+    "e2e": ["{projectRoot}/cypress/**"],
+    "stories": ["{projectRoot}/src/**/*.{?(*([^.]).)stories.{tsx,css},mdx}"],
+    "source": [
+      "{projectRoot}/stencil.config.ts",
+      "{projectRoot}/collection-manifest.json",
+      "{projectRoot}/tailwind.config.js",
+      "{projectRoot}/svg.transform.js",
+      "{workspaceRoot}/tsconfig.json",
+      "{projectRoot}/tsconfig.json",
+      "{projectRoot}/src/**/*([^.]).!({{spec.{ts,tsx},spec.{ts,tsx}.snap},{?(*([^.]).)stories.{tsx,css},mdx}})",
```

Comment:
This line here is a huge mess for two reasons.

1. A named input cannot reference another named input.
    * For instance, I couldn't do `"source": ["!tests", "!stories"]`
2. Negative matchers always have priority
    * For instance, if you have the following named inputs:
      ```json
      {
        "tests": ["{projectRoot}/**/*.spec.ts"],
        "source": [
          "{projectRoot}/**/*.ts",
          "!{projectRoot}/**/*.spec.ts"
        ]
      }
      ```
      and you have a task with the following inputs:
      ```json
      {
        "inputs": ["source", "tests"]
      }
      ```
      Then the `spec.ts` files won't be detected, just as-if you only had `source` as an input.
  * I could be wrong about this, but so far my tests lead me to believe this.

`nx` uses [`fast-glob`](https://github.com/mrmlnc/fast-glob) for matching inputs, so I tested all complex globs like this one to make sure that all correct files are detected.

Reply:
So what happens exactly here if we mess up, forget to update this ?

1- The caching stops being run optimally, and so our build take longer than needed when working locally (not the end of the world) 

2- It will somehow not take new code change into account when it should. So when you work locally, you might get something "corrupted" (for a lack of a better word), that does not actually represent changes you've made locally in a source file ?

If it's 2... that's a big blocker.

Reply:
For this particular line, it would be the first problem: changing some test or story files would cause an unnecessary rebuild of Atomic and its dependants.

However, for other lines than this one, the second problem is a real concern.

If we forget to update inputs to include new file types, then these changes won't be detected, and they might not be rebuilt locally (or even in the CI, if we decide to enable Nx Cloud). This wouldn't necessary be breaking though, since it shouldn't affect the CI, and fixing it should be pretty intuitive.

If we forget to update outputs, then we have a bigger problem. Locally, when loading from the cache, we would randomly get corrupted builds as you said. In the CI, even without Nx Cloud, it would run the first setup job correctly, but then the next jobs would use a corrupted build. This would be a particularly big deal for the publish job(s).

One way to mitigate the problem would be to automatically infer inputs/outputs of each task/target. For example: inferring inputs and outputs from `stencil.ts`. I imagine that's part of why Nx made executors. For example: a Stencil executor should be able to automatically have the correct outputs (AFAIK) by interpreting them from the Stencil configuration file.

Another solution would be to give up on caching. I think caching is going to get increasingly important as we scale the repo by adding more and more projects, but we could skip caching on specific critical jobs (e.g., publishing) or events (e.g., bump commits).

Reply:
Relevant update: https://github.com/coveo/ui-kit/pull/2734#discussion_r1111984473